### PR TITLE
Fix MSSQL getXxxExtendedPropertySQL - fix #4283 issue

### LINKED ISSUE #4283: Doctrine fails to declare Extended Property on SQL-Server array column (when table name is quoted)
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | possibly
| Version     | ORM v2.6.3, DBAL v2.9.2
| Database Server | Microsoft SQL-Server 2017
| Doctrine Driver | `pdo_sqlsrv`

#### Summary

The following error happens when using Doctrine's `SchemaTool` to construct an entity requiring Extended Properties which also has a quoted table name:

```
'An exception occurred while executing 'EXEC sp_addextendedproperty N'MS_Description', N'(DC2Type:array)', N'SCHEMA', 'dbo', N'TABLE', '[the-table]', N'COLUMN', chairs':
                                                                                                                                 
SQLSTATE[42000]: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server] Object is invalid. Extended properties are not permitted on 'dbo.[the-table].chairs', or the object does not exist.'
```

The user has sufficient permissions to add Extended Properties to the database metadata. If I understand this right, this problem is happening because of the quoted table name being passed to the stored procedure as a string (i.e. the databases recognises `[the-table] !== "[the-table]"`.)

Stopping Doctrine from quoting the table name **fixes the problem**, alas I can't afford to rename the table.

#### Current behavior

The database server refuses to execute the query because it does not recognise the referenced object.

#### How to reproduce

1. Declare an entity whose name is a quoted symbol, e.g. 
```php
/**
 *@Table(name="`the-table`")
 */
class TheTable 
{
}
```
2. Add a column requiring the use of Extended Properties by Doctrine

```php
class TheTable
{
 /**
  * @ORM\Column(type="array", name="chairs")
  */
 private $chairs;
}
```

3. Use the Schema Tool to create the tables

```php
$schemaTool = new SchemaTool($entityManager);
$schemaTool->createSchema([
 $entityManager->getMetadataForClass(TheTable::class)
]); 
```

#### Expected behavior

Doctrine can successfully attach the Extended Properties that it needs to the column.

### PR DESCRIPTION
**Pull request by mvorisek**
|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | no
| Fixed issues | #4283

#### Summary

Fix escaping for `EXEC sp_addextendedproperty` SQL function which accept column name as value instead of identifier.

### CODE REVIEW

**Comment by aimeos** [comment]
@mvorisek Any progress on this issue?

**Comment by mvorisek** [comment]
> @mvorisek Any progress on this issue?

PR is RTM from my side

**Comment by aimeos** [comment]
@mvorisek Is there a PR for 3.1.x too?

**Comment by greg0ire** [comment]
@aimeos no need: 2.13.x is regularly merged up in 3.1.x

**Comment by aimeos** [comment]
@morozov @greg0ire  Any chance to merge that PR soon?

**Comment by greg0ire** [comment]
I started reviewing it, then gave up… please rework your commits: one of them reverts what another did, another adds tests, when really there should be one commit with both the tests and the code. Another one mixes questionable style changes with what this PR is all about, making it unnecessarily hard to review it. There is also a merge commit.

IMO there should be 2 commits, maybe 3 but not 6: one for the style changes (dots at the end of line for multiline concatenation, which for some reason not stated, you did not like), and another for the actual fix.

**Comment by aimeos** [comment]
@greg0ire https://github.com/doctrine/dbal/pull/4585/files only shows the real changes that are applied at the end

**Comment by greg0ire** [comment]
@aimeos don't you think I know that?

File: lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php [hunk 1]
```diff
@@ -360,21 +362,18 @@ public function getCreatePrimaryKeySQL(Index $index, $table)
     protected function getCreateColumnCommentSQL($tableName, $columnName, $comment)
     {
         if (strpos($tableName, '.') !== false) {
-            [$schemaSQL, $tableSQL] = explode('.', $tableName);
-            $schemaSQL              = $this->quoteStringLiteral($schemaSQL);
-            $tableSQL               = $this->quoteStringLiteral($tableSQL);
+            [$schemaName, $tableName] = explode('.', $tableName, 2);
```
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php:365]: This implies that a schema name cannot have a dot.

File: lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php [hunk 2]
```diff
@@ -360,21 +362,18 @@ public function getCreatePrimaryKeySQL(Index $index, $table)
...
         } else {
-            $schemaSQL = "'dbo'";
-            $tableSQL  = $this->quoteStringLiteral($tableName);
+            $schemaName = $this->getDefaultSchemaName();
         }
 
         return $this->getAddExtendedPropertySQL(
             'MS_Description',
-            $comment,
+            (string) $comment,
```
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php:372]: Why is this necessary?

File: lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php [hunk 3]
```diff
@@ -796,6 +789,14 @@ protected function getRenameIndexSQL($oldIndexName, Index $index, $tableName)
         ];
     }
 
+    private function quoteSingleIdentifierAsStringLiteral(string $levelName): string
+    {
+        $levelName = preg_replace('~^\[|\]$~s', '', $levelName);
```
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php:794]: Why would one need to pass an explicitly quoted identifier when using SQL Server? Unlike Oracle, quoting doesn't affect the semantics of the identifier.

File: tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL4283Test.php [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+     * @dataProvider columnNameProvider
+     */
+    public function testColumnCommnentOperations(string $columnName): void
+    {
+        if ($this->connection->getDatabasePlatform() instanceof OraclePlatform) {
+            if ($columnName === 'basic') {
+                // Oracle platform does not support other than UC column names
+                $columnName = 'BASIC';
+            } else {
+                $this->markTestSkipped('Oracle platform does not support quoted column names');
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL4283Test.php:46]: What exactly does this mean?

File: tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL4283Test.php [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            }
+
+            // Oracle platform does not support other than UC table names
+            $tableName = 'DBAL4283TBL';
+        } elseif ($this->connection->getDatabasePlatform() instanceof PostgreSqlPlatform) {
+            // PostgreSQL platform does not support quoted table names
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL4283Test.php:52]: What does this mean?

File: tests/Doctrine/Tests/DBAL/Platforms/AbstractSQLServerPlatformTestCase.php [hunk 6]
```diff
@@ -656,7 +656,7 @@ public function testCreateTableWithSchemaColumnComments(): void
         $expectedSql = [
             'CREATE TABLE testschema.test (id INT NOT NULL, PRIMARY KEY (id))',
             "EXEC sp_addextendedproperty N'MS_Description', N'This is a comment', "
-                . "N'SCHEMA', 'testschema', N'TABLE', 'test', N'COLUMN', id",
+                . "N'SCHEMA', 'testschema', N'TABLE', 'test', N'COLUMN', 'id'",
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Platforms/AbstractSQLServerPlatformTestCase.php:659]: The changes from here below seem reasonable.

File: lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php (hunk 1 above)
Comment (mvorisek) [reply lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php:365]: This is how it was, slight improvement when there are more dots, the remaining parts are not silently discarded.

**Comment by mvorisek** [comment]
non-stale
